File: python/src/radar_range_equation/radar_range.py
"""Core radar range equation calculations."""

import math
from typing import Optional


def calculate_max_range(
    transmit_power: float,
    antenna_gain: float,
    wavelength: float,
    radar_cross_section: float,
    min_detectable_signal: float,
) -> float:
    """Calculate the maximum detection range of a radar system.

    Uses the radar range equation to determine the maximum range at which
    a target can be detected.

    Args:
        transmit_power: Transmitted power in Watts
        antenna_gain: Antenna gain (dimensionless, linear scale)
        wavelength: Radar wavelength in meters
        radar_cross_section: Target radar cross section in square meters
        min_detectable_signal: Minimum detectable signal power in Watts

    Returns:
        Maximum range in meters

    Raises:
        ValueError: If any input parameters are non-positive

    Example:
        >>> range_m = calculate_max_range(
        ...     transmit_power=1000,      # 1 kW
        ...     antenna_gain=1000,         # 30 dBi
        ...     wavelength=0.03,           # 10 GHz
        ...     radar_cross_section=1.0,   # 1 m²
        ...     min_detectable_signal=1e-13
        ... )
    """
    if any(x <= 0 for x in [transmit_power, antenna_gain, wavelength, 
                             radar_cross_section, min_detectable_signal]):
        raise ValueError("All parameters must be positive")

    numerator = (transmit_power * antenna_gain**2 * wavelength**2 * 
                 radar_cross_section)
    denominator = (4 * math.pi)**3 * min_detectable_signal
    
    max_range = (numerator / denominator) ** 0.25
    
    return max_range


def calculate_received_power(
    transmit_power: float,
    antenna_gain: float,
    wavelength: float,
    radar_cross_section: float,
    range_m: float,
) -> float:
    """Calculate the received power from a target at a given range.

    Args:
        transmit_power: Transmitted power in Watts
        antenna_gain: Antenna gain (dimensionless, linear scale)
        wavelength: Radar wavelength in meters
        radar_cross_section: Target radar cross section in square meters
        range_m: Range to target in meters

    Returns:
        Received power in Watts

    Raises:
        ValueError: If any input parameters are non-positive

    Example:
        >>> power = calculate_received_power(
        ...     transmit_power=1000,
        ...     antenna_gain=1000,
        ...     wavelength=0.03,
        ...     radar_cross_section=1.0,
        ...     range_m=10000
        ... )
    """
    if any(x <= 0 for x in [transmit_power, antenna_gain, wavelength, 
                             radar_cross_section, range_m]):
        raise ValueError("All parameters must be positive")

    numerator = (transmit_power * antenna_gain**2 * wavelength**2 * 
                 radar_cross_section)
    denominator = (4 * math.pi)**3 * range_m**4
    
    received_power = numerator / denominator
    
    return received_power

File: flutter/lib/src/radar_range.dart
import 'dart:math';

/// Calculate the maximum detection range of a radar system.
///
/// Uses the radar range equation to determine the maximum range at which
/// a target can be detected.
///
/// Parameters:
/// - [transmitPower]: Transmitted power in Watts
/// - [antennaGain]: Antenna gain (dimensionless, linear scale)
/// - [wavelength]: Radar wavelength in meters
/// - [radarCrossSection]: Target radar cross section in square meters
/// - [minDetectableSignal]: Minimum detectable signal power in Watts
///
/// Returns:
/// Maximum range in meters
///
/// Throws [ArgumentError] if any input parameters are non-positive
///
/// Example:
/// ```dart
/// final maxRange = calculateMaxRange(
///   transmitPower: 1000,           // 1 kW
///   antennaGain: 1000,              // 30 dB
///   wavelength: 0.03,               // 10 GHz
///   radarCrossSection: 1.0,         // 1 m²
///   minDetectableSignal: 1e-13,     // -100 dBm
/// );
/// ```
double calculateMaxRange({
  required double transmitPower,
  required double antennaGain,
  required double wavelength,
  required double radarCrossSection,
  required double minDetectableSignal,
}) {
  if (transmitPower <= 0 ||
      antennaGain <= 0 ||
      wavelength <= 0 ||
      radarCrossSection <= 0 ||
      minDetectableSignal <= 0) {
    throw ArgumentError('All parameters must be positive');
  }

  final numerator = transmitPower *
      pow(antennaGain, 2) *
      pow(wavelength, 2) *
      radarCrossSection;
  final denominator = pow(4 * pi, 3) * minDetectableSignal;

  final maxRange = pow(numerator / denominator, 0.25);

  return maxRange as double;
}

/// Calculate the received power from a target at a given range.
///
/// Parameters:
/// - [transmitPower]: Transmitted power in Watts
/// - [antennaGain]: Antenna gain (dimensionless, linear scale)
/// - [wavelength]: Radar wavelength in meters
/// - [radarCrossSection]: Target radar cross section in square meters
/// - [rangeM]: Range to target in meters
///
/// Returns:
/// Received power in Watts
///
/// Throws [ArgumentError] if any input parameters are non-positive
///
/// Example:
/// ```dart
/// final power = calculateReceivedPower(
///   transmitPower: 1000,
///   antennaGain: 1000,
///   wavelength: 0.03,
///   radarCrossSection: 1.0,
///   rangeM: 10000,
/// );
/// ```
double calculateReceivedPower({
  required double transmitPower,
  required double antennaGain,
  required double wavelength,
  required double radarCrossSection,
  required double rangeM,
}) {
  if (transmitPower <= 0 ||
      antennaGain <= 0 ||
      wavelength <= 0 ||
      radarCrossSection <= 0 ||
      rangeM <= 0) {
    throw ArgumentError('All parameters must be positive');
  }

  final numerator = transmitPower *
      pow(antennaGain, 2) *
      pow(wavelength, 2) *
      radarCrossSection;
  final denominator = pow(4 * pi, 3) * pow(rangeM, 4);

  final receivedPower = numerator / denominator;

  return receivedPower as double;
}
